src/app/CommandSender.cpp

```diff
@@ -367,14 +398,47 @@ CHIP_ERROR CommandSender::ProcessInvokeResponseIB(InvokeResponseIB::Parser & aIn
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CommandSender::PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct)
+CHIP_ERROR CommandSender::GetAdditionalInvokeResponseElements(AdditionalInvokeResponseElements & aInvokeReponseElements)
+{
+    VerifyOrReturnError(mState == State::ResponseReceived, CHIP_ERROR_INCORRECT_STATE);
+    aInvokeReponseElements = mAdditionalResponseElements;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommandSender::SetBatchCommandsConfig(uint16_t aRemoteMaxPathsPerInvoke)
+{
+#if CHIP_CONFIG_SENDING_BATCH_COMMANDS_ENABLED
+    VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(aRemoteMaxPathsPerInvoke, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mRemoteMaxPathsPerInvoke = aRemoteMaxPathsPerInvoke;
+    mBatchCommandsEnabled    = (aRemoteMaxPathsPerInvoke > 1);
+    return CHIP_NO_ERROR;
+#else
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+#endif
+}
+
+CHIP_ERROR CommandSender::PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct,
+                                         AdditionalCommandDataElements aOptionalArgs)
 {
     ReturnErrorOnFailure(AllocateBuffer());
 
     //
-    // We must not be in the middle of preparing a command, or having prepared or sent one.
+    // We must not be in the middle of preparing a command, or have already sent InvokeReponseMessage.
     //
-    VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
+    bool canAddAnotherCommand = mState == State::AddedCommand && mBatchCommandsEnabled;
+    VerifyOrReturnError(mState == State::Idle || canAddAnotherCommand, CHIP_ERROR_INCORRECT_STATE);
+    // TODO what is the best error to give here? I thought CHIP_ERROR_SENTINEL might be an indicator for the called to then
```

Comment:
we can create a brand new error specifically for this case. a `CHIP_ERROR_MAXIMUM_PATHS_PER_INVOKE_EXCEEDED` would be very specific in case we require explicit processing for this error case. `Sentinel` is not that intuitive to me. If we don't need specific handling some `out of resources` (like no memory) is used in other cases.